src/TelemetryConsumers/Orleans.TelemetryConsumers.Linux/LinuxEnvironmentStatistics.cs

```diff
@@ -0,0 +1,276 @@
+
+            _cts = new CancellationTokenSource();
+            ct.Register(() => _cts.Cancel());
+
+            _monitorTask = await Task.Factory.StartNew(
+                () => Monitor(_cts.Token),
+                _cts.Token,
+                TaskCreationOptions.DenyChildAttach | TaskCreationOptions.RunContinuationsAsynchronously,
+                TaskScheduler.Default
+            );
+
+            _logger.LogTrace($"Started {nameof(LinuxEnvironmentStatistics)}");
+        }
+
+        public async Task OnStop(CancellationToken ct)
+        {
+            if (_cts == null)
+                return;
+
+            _logger.LogTrace($"Stopping {nameof(LinuxEnvironmentStatistics)}");
+            try
+            {
+                _cts.Cancel();
+                try
+                {
+                    await _monitorTask;
+                }
+                catch (TaskCanceledException) { }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, $"Error stopping {nameof(LinuxEnvironmentStatistics)}");
+            }
+            finally
+            {
+                _logger.LogTrace($"Stopped {nameof(LinuxEnvironmentStatistics)}");
+            }
+        }
+
+        private async Task UpdateTotalPhysicalMemory()
+        {
+            var memTotalLine = await ReadLineStartingWithAsync("/proc/meminfo", "MemTotal");
+
+            if (string.IsNullOrWhiteSpace(memTotalLine))
+            {
+                _logger.LogWarning("Couldn't read 'MemTotal' line from '/proc/meminfo'");
+                return;
+            }
+
+            // Format: "MemTotal:       16426476 kB"
+            if (!long.TryParse(new string(memTotalLine.Where(char.IsDigit).ToArray()), out var totalMemInKb))
+            {
+                _logger.LogWarning($"Couldn't parse meminfo output");
+                return;
+            }
+
+            TotalPhysicalMemory = totalMemInKb * 1_000;
+        }
+
+        private long _prevIdleTime;
+        private long _prevTotalTime;
+
+        private async Task UpdateCpuUsage(int i)
+        {
+            var cpuUsageLine = await ReadLineStartingWithAsync("/proc/stat", "cpu  ");
+
+            if (string.IsNullOrWhiteSpace(cpuUsageLine))
+            {
+                _logger.LogWarning("Couldn't read line from '/proc/stat'");
+                return;
+            }
+
+            // Format: "cpu  20546715 4367 11631326 215282964 96602 0 584080 0 0 0"
+            var cpuNumberStrings = cpuUsageLine.Split(' ').Skip(2);
+
+            if (cpuNumberStrings.Any(n => !long.TryParse(n, out _)))
+            {
+                _logger.LogWarning($"Failed to parse '/proc/stat' output correctly. Line: {cpuUsageLine}");
+                return;
+            }
+
+            var cpuNumbers = cpuNumberStrings.Select(long.Parse).ToArray();
+            var idleTime = cpuNumbers[3];
+            var totalTime = cpuNumbers.Sum();
+
+            if (i > 0)
```

Comment:
If for some reason it fails at the first call, `i` will be superior to `0` and you might report some wirted data. I don't think it is a huge issue.